baxter_collaboration_lib/src/robot_interface/robot_interface.cpp

```diff
@@ -103,26 +103,42 @@ RobotInterface::RobotInterface(string _name, string _limb, bool _use_robot, doub
 
 bool RobotInterface::startThread()
 {
-    return ctrl_thread.start(ThreadEntryFunc, this);
+    // return ctrl_thread.start(ThreadEntryFunc, this);
+    ctrl_thread = std::thread(ThreadEntryFunc, this);
+	return ctrl_thread.joinable(); // std::thread::joinable checks if thread identifies as active thread of execution
 }
 
+/**
 bool RobotInterface::closeThread()
 {
     return ctrl_thread.close();
 }
 
 bool RobotInterface::killThread()
 {
-    return ctrl_thread.kill();
+    // return ctrl_thread.kill();
+}
+**/
+
+void RobotInterface::setThreadKill(bool arg)
+{
+    std::lock_guard<std::mutex> lck(_mtx_thread_kill);
+    _thread_kill = arg;
+}
+
+bool RobotInterface::getThreadKill()
+{
+    std::lock_guard<std::mutex> lck(_mtx_thread_kill);
+    return _thread_kill;
 }
 
 void RobotInterface::ThreadEntry()
 {
-    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+    // pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
 
     ros::Rate r(ctrl_freq);
 
-    while (RobotInterface::ok())
+    while ((not ros::isShuttingDown()) && (getThreadKill() == false))
```

Comment:
maybe now is better to move `not ros::isShuttingDown()` to `ros::ok()` (test please)